### PR TITLE
sql-assessment extension code

### PR DESCRIPTION
**Pull request by vchvlad**
This is a sql-assessment extension that has been built on model view components.

### CODE REVIEW

File: extensions/sql-assessment/src/engine.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+import * as azdata from 'azdata';
+import { createHistoryFileName, readHistoryFileNames, getAssessmentDate, TargetWithChildren } from './utils';
+import { promises as fs } from 'fs';
+
+export enum AssessmentType {
+	AvailableRules = 1,
+	InvokeAssessment = 2
+}
+
+export type OnResultCallback = (result: azdata.SqlAssessmentResult, assessmentType: AssessmentType, append: boolean) => any;
```
Comment (Charles-Gagnon) [inline extensions/sql-assessment/src/engine.ts:16]: no better return type than any?

File: extensions/sql-assessment/src/engine.ts [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	private async loadHistory(): Promise<void> {
+		this.historicalRecords = [];
+		const files: TargetWithChildren[] = await readHistoryFileNames(this.targetName);
+
+		for (let nFileName = 0; nFileName < files.length; nFileName++) {
+			const file: TargetWithChildren = files[nFileName];
+			const content: string = await fs.readFile(file.target, 'utf8');
+			const result: azdata.SqlAssessmentResult = JSON.parse(content);
+
+			for (let nChild = 0; nChild < file.children.length && this.isServerConnection; nChild++) {
```
Comment (Charles-Gagnon) [inline extensions/sql-assessment/src/engine.ts:114]: Seems odd to check isServerConnection each iteration - is that ever going to change mid-execution?

File: extensions/sql-assessment/src/htmlReportGenerator.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+			</div>
+			${this.buildStyleSection()}
+		</body>
+		</html>`;
+		return mainContent;
+	}
+
+	private instanceName(serverInfo: azdata.ServerInfo): string {
+		const serverName = this._connectionInfo.serverName;
+		if (['local', '(local)', '(local);'].indexOf(serverName.toLowerCase()) >= 0) {
```
Comment (Charles-Gagnon) [inline extensions/sql-assessment/src/htmlReportGenerator.ts:43]: Is the ; next to (local) on purpose?

File: extensions/sql-assessment/src/htmlReportGenerator.ts [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+			return serverInfo !== undefined
+				? (<any>serverInfo)['machineName']
+				: serverName;
+		}
+		return serverName;
+	}
+
+	private buildVersionDetails(serverInfo: azdata.ServerInfo): string {
```
Comment (Charles-Gagnon) [inline extensions/sql-assessment/src/htmlReportGenerator.ts:52]: It seems like this isn't being used?

File: extensions/sql-assessment/src/maincontroller.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		this.sqlAssessment = ((await vscode.extensions.getExtension(mssql.extension.name)?.activate() as mssql.IExtension)).sqlAssessment;
+		this.engine = new AssessmentEngine(this.sqlAssessment);
+		this.registerModelView();
+		//TelemetryReporter.sendViewEvent(SqlAssessmentTelemetryView);
+		return true;
+	}
+
+
+
+	private registerModelView(): void {
```
Comment (Charles-Gagnon) [inline extensions/sql-assessment/src/maincontroller.ts:43]: registerModelViewProvider

File: extensions/sql-assessment/src/telemetry.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (Charles-Gagnon) [inline extensions/sql-assessment/src/telemetry.ts:1]: Remove this stuff and the package if you're not using it currently

File: extensions/sql-assessment/src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+		Number(strDate.substr(12, 2)) // s
+	);
+	return date.getTime() - date.getTimezoneOffset() * 60 * 1000;
+}
+
+function generateDefaultFileName(resultDate: Date): string {
+	return `${resultDate.toISOString().replace(/-/g, '').replace('T', '').replace(/:/g, '').split('.')[0]}`;
+}
+
+export function escape(html: string): string {
```
Comment (Charles-Gagnon) [inline extensions/sql-assessment/src/utils.ts:90]: Suggest htmlEscape for the name so it's clear what type of escaping is being done

**Review by Charles-Gagnon — APPROVED** [review]

File: extensions/sql-assessment/src/htmlReportGenerator.ts (hunk 4 above)
Comment (vchvlad) [reply extensions/sql-assessment/src/htmlReportGenerator.ts:52]: This is needed for the header to display information about the SQL server instance.

File: extensions/sql-assessment/src/telemetry.ts (hunk 6 above)
Comment (vchvlad) [reply extensions/sql-assessment/src/telemetry.ts:1]: I have brought this back.

**Pull request merged**